cantor-base/src/main/java/com/salesforce/cantor/Events.java

```diff
@@ -256,12 +256,163 @@ default void store(String namespace,
+    }
+
+    /**
+     * Get all events in the given namespace, with timestamp between the start and end, and metadata and dimensions
+     * matching the given queries.
+     *
+     * A metadata query can match against an exact value for a key, or use wild-card
+     * character (i.e., '*'). For example: "host" => "localhost" matches with all events where the value of metadata
+     * with key "host" is exactly "localhost"; and the query "host" => "~prod-*-example" matches all events where the
+     * value is "like 'prod-*-example'". Note that wild-card queries have to start with "~", exact queries can
+     * optionally start with "=".
+     *
+     * A dimensions query can match against exact values or less-than, less-than-or-equal, more-than, more-than-or-equal
+     * values for a set of dimensions. For example: "cpu" => ">=0.3" matches all events where the dimension with key
+     * "cpu" has a value higher than or equal to 0.3.
+     *
+     * @param namespace the namespace identifier
+     * @param startTimestampMillis start UTC timestamp in milli-seconds
+     * @param endTimestampMillis end UTC timestamp in milli-seconds
+     * @param metadataQuery map of string to string representing a query to run against events metadata
+     * @param dimensionsQuery map of string to string representing a query to run against events dimensions
+     * @param includePayloads flag to include payloads in the response or not; if false, event.getPayload() returns null
+     * @param ascending order results ascending if true; descending if false
+     * @param limit maximum number of events to return
+     * @return list of all events in the namespace with timestamp between start/end
+     * and metadata/dimensions matching the query
+     * @throws IOException exception thrown from the underlying storage implementation
+     */
     List<Event> get(String namespace,
                     long startTimestampMillis,
                     long endTimestampMillis,
                     Map<String, String> metadataQuery,
                     Map<String, String> dimensionsQuery,
-                    boolean includePayloads) throws IOException;
+                    boolean includePayloads,
+                    boolean ascending,
+                    int limit) throws IOException;
+
+    /**
+     * Get the first (earliest) event in the given namespace, with timestamp between the start and end.
+     *
+     * @param namespace the namespace identifier
+     * @param startTimestampMillis start UTC timestamp in milli-seconds
+     * @param endTimestampMillis end UTC timestamp in milli-seconds
+     * @return the earliest event in the namespace with timestamp between start/end and metadata/dimensions matching the query
+     * @throws IOException exception thrown from the underlying storage implementation
+     */
+    default List<Event> first(String namespace,
```

Comment:
Should we return just `Event`? Since it's only ever a single event (or nothing). Same question for `last`